pkg/syncer/syncers/downsyncer.go

```diff
@@ -0,0 +1,181 @@
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+)
+
+type DownSyncer struct {
+	logger                  klog.Logger
+	upstreamClientFactory   ClientFactory
+	downstreamClientFactory ClientFactory
+	upstreamClients         map[schema.GroupKind]*Client
+	downstreamClients       map[schema.GroupKind]*Client
+}
+
+func NewDownSyncer(logger klog.Logger, upstreamClientFactory ClientFactory, downstreamClientFactory ClientFactory, syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) (*DownSyncer, error) {
+
+	downSyncer := DownSyncer{
+		logger:                  logger,
+		upstreamClientFactory:   upstreamClientFactory,
+		downstreamClientFactory: downstreamClientFactory,
+	}
+
+	if err := downSyncer.initializeClients(syncedResources, conversions); err != nil {
+		logger.Error(err, "failed to initialize clients")
+		return &downSyncer, err
+	}
+
+	return &downSyncer, nil
+}
+
+func (ds *DownSyncer) initializeClients(syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.upstreamClients = map[schema.GroupKind]*Client{}
+	ds.downstreamClients = map[schema.GroupKind]*Client{}
+	return initializeClients(ds.logger, syncedResources, ds.upstreamClientFactory, ds.downstreamClientFactory, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) ReInitializeClients(syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	return initializeClients(ds.logger, syncedResources, ds.upstreamClientFactory, ds.downstreamClientFactory, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) getClients(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) (*Client, *Client, error) {
+	return getClients(resource, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) SyncOne(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.logger.V(3).Info(fmt.Sprintf("downsync '%s'", resourceToString(resource)))
+	upstreamClient, downstreamClient, err := ds.getClients(resource, conversions)
+	if err != nil {
+		ds.logger.Error(err, fmt.Sprintf("failed to get client '%s'", resourceToString(resource)))
+		return err
+	}
+	resourceForUp := convertToUpstream(resource, conversions)
+	ds.logger.V(3).Info(fmt.Sprintf("  get '%s' from upstream", resourceToString(resourceForUp)))
+	upstreamResource, err := upstreamClient.Get(resourceForUp)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			ds.logger.V(3).Info(fmt.Sprintf("  not found '%s' in upstream", resourceToString(resourceForUp)))
+			ds.logger.V(3).Info(fmt.Sprintf("  skip downsync '%s'", resourceToString(resourceForUp)))
```

Comment:
Should there be a downstream delete here?

Reply:
Yes, I will create a gitissue for deletion support.

Reply:
I will create a gitissue for deletion support.